### PR TITLE
#22119 Moving window throttler

### PR DESCRIPTION
**Pull request by VRanga000**

### CODE REVIEW

**Comment by typesafehub-validator** [comment]
Hi @VRanga000,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

**Comment by ktoso** [comment]
Please link back from PRs to the issue it is addressing so we know that it's meant to solve a given issue. 
Refs https://github.com/akka/akka/pull/22121

**Comment by ktoso** [comment]
OK TO TEST

**Comment by johanandren** [comment]
Hi @VRanga000 we would need you to sign the Lightbend CLA to be able to accept this as described by the comment by the CLA bot earlier. Please let us know if you are not ok with that (in which case I'm afraid we cannot accept this PR). Thanks!

**Comment by VRanga000** [comment]
Thanks folks for seeing this thru, unfortunately I am still waiting for the relevant group at my company to approve/accept the CLA.

**Comment by patriknw** [comment]
Refs #22119

**Comment by patriknw** [comment]
@VRanga000 I'm not sure we should put more things into the contrib/throttle ?
Pretty much the same thing can be done with 3 lines of Akka Streams:

```
val throttle: RunnableGraph[ActorRef] =
    Source.actorRef(bufferSize = 1000, OverflowStrategy.dropNew)
      .throttle(100, 1.second, 10, ThrottleMode.Shaping)
      .to(Sink.actorRef(target, NotUsed))
```

I'm not sure the Akka streams `throttle` stage supports what you have done here. Could you investigate that?

**Comment by raboof** [comment]
As `akka-contrib` is being deprecated perhaps this should move into a standalone repository?

**Comment by patriknw** [comment]
@VRanga000 akka-contrib is deprecated but if you would like to have this feature included we can merge it after rebase, otherwise we should close this PR. `throttle` in Akka Streams is the replacement http://doc.akka.io/docs/akka/current/scala/project/migration-guide-2.4.x-2.5.x.html#timerbasedthrottler

**Comment by VRanga000** [comment]
Lightbend CLA has been (finally) accepted. I leave it to you to close this PR or merge it (we are currently using this implementation internally). The streams option does seem like the way to go forward.

File: akka-contrib/src/main/scala/akka/contrib/throttle/TimerBasedThrottler.scala [hunk 1]
```diff
@@ -295,10 +300,30 @@ class TimerBasedThrottler(var rate: Rate) extends Actor with FSM[State, Data] {
    */
   private def deliverMessages(data: Data): Data = {
     val queue = data.queue
-    val nrOfMsgToSend = scala.math.min(queue.length, data.callsLeftInThisPeriod)
 
-    queue.take(nrOfMsgToSend).foreach(x ⇒ data.target.get.tell(x.message, x.sender))
+    import akka.pattern.ask
+    implicit val timeout = Timeout(61.seconds)
+    val callsInThisPeriod =
+      Try(Await.result((stopwatch ? LastPeriodCount(rate.duration)).mapTo[LastPeriodCountResponse], 60.seconds)) match {
```
Comment (patriknw) [inline akka-contrib/src/main/scala/akka/contrib/throttle/TimerBasedThrottler.scala]: Blocking Await is not nice here, I think we'll close this without merging. Thanks anyway. Shouldn't be a problem to maintain a copy of this in your project.

**Comment by patriknw** [comment]
Refs #22119

**Pull request closed without merging**